### PR TITLE
Project: allow connecting a project to a remote repository after it has been created

### PR DESCRIPTION
**Pull request by stsewd**
- Several methods were moved from the ProjectBasicsForm class to the ProjectForm class, and UpdateProjectForm now inherits from ProjectForm. This was so we can use the default form field of the remote repository foreign key, while also sharing some code.

vlcsnap-2024-07-24-17h03m38s272

image

Closes https://github.com/readthedocs/readthedocs.org/issues/9437

### CODE REVIEW

**Comment by agjohnson** [comment]
I have been pushing to use verbose text for the default/initial widget text:

image

If a project has a connected repository, it should be selected in this dropdown, correct? If so, when there is no connected repository the dropdown initial value might be "No connected repository" instead of "-------".

**Comment by stsewd** [comment]
> If a project has a connected repository, it should be selected in this dropdown, correct? 

yep, I'll change the text of the default empty selection.

**Comment by agjohnson** [comment]
It might also be worth using something like "Connected repository" instead of "Remote repository" in the user UI, to give a stronger framing to the user about the purpose of the field -- "remote repository" doesn't quite do this.

@stsewd Do the view changes affect the project creation views on either dashboard at all? I'm having trouble figuring this out by looking at the code.

File: readthedocs/projects/forms.py [hunk 1]
```diff
@@ -52,13 +52,61 @@ def __init__(self, *args, **kwargs):
...
 
+        self.fields["repo"].widget.attrs["placeholder"] = self.placehold_repo()
+        self.fields["repo"].widget.attrs["required"] = True
+
+        queryset = RemoteRepository.objects.for_project_linking(self.user)
+        current_remote_repo = (
+            self.instance.remote_repository if self.instance.pk else None
+        )
+        # If there is a remote repo attached to the project, add it to the queryset,
+        # since the current user might not have access to it.
```
Comment (agjohnson) [inline readthedocs/projects/forms.py:63]: Good call :+1:

**Comment by stsewd** [comment]
> @stsewd Do the view changes affect the project creation views on either dashboard at all? I'm having trouble figuring this out by looking at the code.

Nope, that input is still hidden from the creation form.

File: readthedocs/rtd_tests/tests/test_project_forms.py [hunk 2]
```diff
@@ -333,6 +335,55 @@ def test_trigger_build_on_save(self, trigger_build):
...
+
+        # No remote repository attached.
+        form = UpdateProjectForm(data, instance=self.project, user=self.user)
+        self.assertTrue(form.is_valid())
+
+        # Remote repository attached, but it doesn't belong to the user.
+        data["remote_repository"] = remote_repository.pk
+        form = UpdateProjectForm(data, instance=self.project, user=self.user)
+        self.assertFalse(form.is_valid())
+        self.assertIn("remote_repository", form.errors)
```
Comment (agjohnson) [inline readthedocs/rtd_tests/tests/test_project_forms.py:365]: Ah so, you added the addition in the widget query for showing the remote repo even if the user doesn't have access to it, but it seems this user might be blocked from updating the form at all in this case too?

Like, if a user without access to the project's remote repository wants to change another field in the form?

File: readthedocs/rtd_tests/tests/test_project_forms.py [hunk 3]
```diff
@@ -333,6 +335,55 @@ def test_trigger_build_on_save(self, trigger_build):
...
+        self.assertTrue(form.is_valid())
+
+        # The project has the remote repository attached.
+        # And the user doesn't have access to it anymore, but still can use it.
+        self.project.remote_repository = remote_repository
+        self.project.save()
+        remote_repository_rel.delete()
+        data["remote_repository"] = remote_repository.pk
+        form = UpdateProjectForm(data, instance=self.project, user=self.user)
+        self.assertTrue(form.is_valid())
```
Comment (agjohnson) [inline readthedocs/rtd_tests/tests/test_project_forms.py:385]: Or am I wrong above? This is maybe the case I'm considering.

**Review by agjohnson — COMMENTED** [review]
A couple more notes. I think this is looking really good, I had a couple other comments on the PR that might change this a little bit.

File: readthedocs/rtd_tests/tests/test_project_forms.py (hunk 2 above)
Comment (stsewd) [reply readthedocs/rtd_tests/tests/test_project_forms.py:365]: > Like, if a user without access to the project's remote repository wants to change another field in the form?

No, the current remote repository is included in the queryset itself, so it's an allowed value when saving the form.

**Review by agjohnson — APPROVED** [review]
This looks great!

After this as a first pass, I think we'll eventually want a polish pass on this that adds some more UI or maybe just moves the form view too. This is mostly to add a modal warning or something "Hey, don't change this if you're not sure exactly what this is doing".

But also, we talked about making this a nice UI too, so perhaps that's all one project and it all belongs in a dedicated admin UI.

**Comment by humitos** [comment]
Is it possible to make the field searchable in the new dashboard? (similar to what we do at "Add project"). Users with a lot of repositories will have issues finding the correct one.

**Comment by stsewd** [comment]
> Is it possible to make the field searchable in the new dashboard? (similar to what we do at "Add project"). Users with a lot of repositories will have issues finding the correct one.

It's searchable

**Pull request merged**